hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/TemplatePathsTraverser.java

```diff
@@ -132,252 +141,267 @@ public Traverser(List<RepeatEdgeStep> steps,
-            checkCapacity(this.capacity, this.accessedNodes(),
-                          "template paths");
-            if (this.limit == NO_LIMIT || this.pathCount() < this.limit) {
-                return false;
-            }
-            return true;
+        public abstract void traverseOneLayer(
+                             Map<Id, List<Node>> vertices,
+                             RepeatEdgeStep step,
+                             BiConsumer<Id, RepeatEdgeStep> consumer);
+
+        public void afterTraverse(RepeatEdgeStep step, boolean forward) {
+            Map<Id, List<Node>> all = forward ? this.sourcesAll :
+                                                this.targetsAll;
+            this.addNewVerticesToAll(all);
+            this.reInitCurrentIfNeeded(step, forward);
+            this.stepCount++;
         }
 
-        public boolean lastSuperStep() {
-            return this.targetIndex == this.sourceIndex ||
-                   this.targetIndex == this.sourceIndex + 1;
+        private void forward(Id v, RepeatEdgeStep step) {
+            this.traverseOne(v, step, true);
         }
-    }
 
-    private class ConcurrentTraverser extends Traverser {
+        private void backward(Id v, RepeatEdgeStep step) {
+            this.traverseOne(v, step, false);
+        }
 
-        private ConcurrentMultiValuedMap<Id, Node> sources =
-                new ConcurrentMultiValuedMap<>();
-        private ConcurrentMultiValuedMap<Id, Node> targets =
-                new ConcurrentMultiValuedMap<>();
-        private ConcurrentMultiValuedMap<Id, Node> sourcesAll =
-                new ConcurrentMultiValuedMap<>();
-        private ConcurrentMultiValuedMap<Id, Node> targetsAll =
-                new ConcurrentMultiValuedMap<>();
+        private void traverseOne(Id v, RepeatEdgeStep step, boolean forward) {
+            if (this.reachLimit()) {
+                return;
+            }
 
-        private Set<Path> paths;
+            Iterator<Edge> edges = edgesOfVertex(v, step);
+            while (edges.hasNext()) {
+                HugeEdge edge = (HugeEdge) edges.next();
+                Id target = edge.id().otherVertexId();
 
-        public ConcurrentTraverser(Collection<Id> sources,
-                                   Collection<Id> targets,
-                                   List<RepeatEdgeStep> steps, boolean withRing,
-                                   long capacity, long limit) {
-            super(steps, capacity, limit, withRing);
-            for (Id id : sources) {
-                this.sources.add(id, new Node(id));
+                this.processOne(v, target, forward);
             }
-            for (Id id : targets) {
-                this.targets.add(id, new Node(id));
+        }
+
+        private void processOne(Id source, Id target, boolean forward) {
+            if (forward) {
+                processOneForForward(source, target);
+            } else {
+                processOneForBackward(source, target);
```

Comment:
can we merge the two methods with one, and pass sourcesAll or targetsAll to that method